falcon/api.py

```diff
@@ -556,7 +558,10 @@ def handle(req, resp, ex, params):
 
         # Insert at the head of the list in case we get duplicate
         # adds (will cause the most recently added one to win).
-        self._error_handlers.insert(0, (exception, handler))
```

Comment:
Thanks for your comments @vytas7. Well, the idea of using EAFP stemmed from the fact that I cannot say `issubclass((ValueError, TypeError), BaseException)` (arg 1 must be a class), so in some way I have to first check if what the user has passed to the function is an iterable or not. AFAIK, a simple, pythonic way to do that is EAFP. 

I agree with you that it would be better to have some stricter checks on the nature of the `exception` argument. One way to go could be first simply normalize to a tuple with EAFP, and then check `all(issubclass(exc, BaseException) for exc in exception_tuple)`: if yes, I would add them to the list of handlers, otherwise, raise a descriptive `TypeError`. What do you think?

Reply:
> arg 1 must be a class
Ouch, I should have tested that before suggesting... :man_facepalming: 

But maybe we could still build upon that idea?
```python
if isinstance(exception, type) and issubclass(exception, BaseException):
    self._error_handlers.insert(0, (exception, handler))
else:
    self._error_handlers.insert(0, (tuple(exception), handler))
```

I think I would prefer this form. But I realize it is much about stylistic preferences, and not necessarily a huge difference in logic.

Reply:
so, the point here is that we want to allow `exception` to be a generic iterable of exceptions, or a single exception, that can be made into a tuple which can then be used by `_handle_error()` to check against the currently occurring exception. 

We cannot use `isinstance` or `issubclass` to test for "cast-ability" to a tuple, so EAFP would be the way to go for this. The we can also check that `exception` (or each element of `exception`) is indeed a subclass of `BaseException`. Give me 10min and I'll try to show you the code..

Reply:
the alternative to simplify things is to allow only single exceptions or tuples, and document this. I actually started this PR because in one case I found it useful to accumulate/remove some exceptions into a list and pass it directly to the `add_error_handler`. I'm available to both options though, I have no strong opinions about it.

Reply:
In this case I think EAFP is justified, given that either way the only way to reliably test for an iterable is to attempt it and see if it raises an exception. If passing a single exception were the most common use case, and this logic were on a hot code path, checking the type up front would be ideal. But under the circumstances that's a moot point, so I'd be cool with either of the proposed approaches.

Reply:
I'm not a fan of EAFP, but I see here it allows to continue to part two in an elegant fashion (my previous comment precedes the latest commit), so I'm fine with it.

I just a have a small stylistic nitpick, and would like to add one more test for the `TypeError` case, and I'm fine to approve this :+1: